rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -336,8 +344,46 @@ object Reduce {
         case _ => Task raiseError new Error("Unimplemented expression.")
       }
 
-    def eval(mat: Match)(implicit env: Env[Par]): Task[Unit] =
-      Task raiseError new Error("match is currently unimplemented.")
+    def eval(mat: Match)(implicit env: Env[Par]): Task[Unit] = {
+      def addToEnv(env: Env[Par], freeMap: Map[Int, Par], freeCount: Int): Env[Par] =
+        Range(0, freeCount).foldLeft(env)(
+          (acc, e) =>
+            acc.put(
+              freeMap.get(e) match {
+                case Some(p) => p
+                case None    => Par()
+              }
+          )
+        )
+      @tailrec
+      def firstMatch(target: Par, cases: Seq[MatchCase]): Task[Unit] =
+        cases match {
+          case Nil => Task.unit
+          case singleCase +: caseRem =>
+            val matchResult = SpatialMatcher
+              .spatialMatch(target, singleCase.pattern.get)
+              .runS(SpatialMatcher.emptyMap)
+            matchResult match {
+              case None => firstMatch(target, caseRem)
+              case Some(freeMap) => {
+                println("freeMap is: " + freeMap)
+                val newEnv = addToEnv(env, freeMap, singleCase.pattern.get.freeCount)
+                println("freeCount is: " + singleCase.pattern.get.freeCount)
+                println("original env is: " + env)
+                println("newEnv is: " + newEnv)
```

Comment:
Do you want to keep these print statements for the future?

Reply:
No. They're removed.